echo/echo.1
.\"	$NetBSD: echo.1,v 1.13 2003/08/07 09:05:12 agc Exp $
.\"
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)echo.1	8.1 (Berkeley) 7/22/93
.\"
.Dd July 22, 1993
.Dt ECHO 1
.Os
.Sh NAME
.Nm echo
.Nd write arguments to the standard output
.Sh SYNOPSIS
.Nm
.Op Fl n
.Op Ar string ...
.Sh DESCRIPTION
The
.Nm
utility writes any specified operands, separated by single blank (`` '')
characters and followed by a newline (``\en'') character, to the standard
output.
.Pp
The following option is available:
.Bl -tag -width flag
.It Fl n
Do not print the trailing newline character.
.El
.Sh EXIT STATUS
.Ex -std echo
.Sh SEE ALSO
.Xr printf 1
.Sh STANDARDS
The
.Nm
utility is expected to be
.St -p1003.2
compatible.


doc/NEWS
======================================================================
Version 8.0.0.1, by Aleksey Cheusov, Sat, 27 Mar 2021 23:18:53 +0300

 When LINKSPREFIX is set:
   - programs prefix if "nb" instead of "nb-"
   - MANZ is handled correctly

======================================================================
Version 7.0.0.11, by Aleksey Cheusov, Mon, 22 Mar 2021 11:21:43 +0300

 Enable "ul" and "tabs" on Linux/glibc.

 Enable "cmp" and "find" on systems without strtoq(3).

 Fixes for nawk, mtree, csplit, find, tabs, ul, cal, cut.

 Fixes for following features: parsedate, bsd_signal, curses.

 Minimize patch size for a lot of utilities.

======================================================================
Version 7.0.0.10, by Aleksey Cheusov, Fri, 27 Nov 2020 15:22:51 +0300

  Disable grep(1) if REG_STARTEND is not defined in regex.h.

  chmod, cp, du, find, ls, mtree, pax, rm: improvements on systems
  without fts_open(3) functions family (e.g., musl libc and Solaris).

  Improve portability of xinstall(1) and mtree(1).

======================================================================
Version 7.0.0.9, by Aleksey Cheusov, Thu, 14 May 2020 20:46:44 +0300

  Fix some build failures on Linux-es caused by __wur
  and '-O2 -D_FORTIFY_SOURCE=2'

======================================================================
Version 7.0.0.8, by Aleksey Cheusov, Mon, 16 Mar 2020 01:38:17 +0300

  Add one line description to projects.
  So, one can run "mkcmake help"

======================================================================
Version 7.0.0.7, by Aleksey Cheusov, Thu, 20 Feb 2020 01:10:57 +0300

  Requires mk-configure 0.34.0 or above to build

  Remove broken hack with CC_TYPE, latest mk-configure is able to do
  everything automatically

  Use new mk-configure's feature libl instead of LEXLIB variable

  Use CFLAGS.check for disabling warnerr

  mknod: fix build failure on latest OpenSuSE tumbleweed

  id, date: fixes for Darwin-15.6.0 (disabling warnings)

======================================================================
Version 7.0.0.6, by Aleksey Cheusov, Sun, 19 Jan 2020 01:45:59 +0300

  Add support for Solaris-10 and -11 to a lot of utilities
  (almost all with the help of musl-fts library!).

  A lot of fixes for SunC compiler (versions 5.11 and 5.15).

  mk/mkc_imp.f_bsd_getopt.mk: fix for target "clean"

  Fix feature "efun" on SunOS-5.10.

  Add new feature "strsep" that provides portable strsep(3) and stresep(3)

  find, ls, mtree and xinstall: fix build failure on eglibc-2.13

======================================================================
Version 7.0.0.5, by Aleksey Cheusov, Mon,  6 Jan 2020 01:43:19 +0300

  Fix typos doc/INSTALL

  Fix bugs in the local feature "base64". As a result we fix some
  compilation issues on musl-based Linuxes (Alpine Linux).

  Fix compilation failures (seen on musl-based Alpine Linux) of the
  following programs: compress, sort, hexdump, id, leave, uudecode

  Implement new local feature "bsd_signal" that implements
  BSD functions "sigblock", "sigsetmask", "sigmask" and "siggetmask".

======================================================================
Version 7.0.0.4, by Aleksey Cheusov, Wed,  1 May 2019 14:40:14 +0300

  A lot of fixes and improvements from previous release in utilities
  and portability code.
  See https://github.com/cheusov/nbase for details.

  Introduce CURSES_LIBNAME build-time variable (see doc/INSTALL)

  Almost all utilities set WARNERR to YES.

  Latest mk-configure is needed for build.

  This release was successfully tested on SunOS-5.11, FreeBSD-12.0.3,
  NetBSD-8 and 8.99, OpenBSD-6.4, Darwin and diverse glibc-based
  Linuxes. GCC and CLang compilers were tested.

======================================================================
Version 7.0.0.3, by Aleksey Cheusov, Thu,  8 Mar 2018 14:51:38 +0300

  "install", "cksum", "mknod" and "mtree" were added to the
  collection.

  All utilities were ported to OpenBSD and Darwin.  They were
  successfully built on Darwin-15.6.0/amd64, Linux(glibc)/amd64 and
  ARMv7, NetBSD-current/amd64 and OpenBSD-5.3/i386.

  All utilities are successfully build using gcc and clang.

  nl: fix for glibc-based systems (Linux).

  Fixes for Out-of-tree build failures.

  mk-configure 0.30.0 or above is needed for build.

  Fixes for helper scripts (helpers/*)

  A lot of compilation warnings fixes.

======================================================================
Version 7.0.0.2, by Aleksey Cheusov, Mon, 10 Jul 2017 00:37:48 +0300

  Initial release


echo/echo.c
/* $NetBSD: echo.c,v 1.18 2008/09/18 05:42:08 dholland Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "mkc_progname.h"
#include "mkc_posix_getopt.h"

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT(
"@(#) Copyright (c) 1989, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)echo.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: echo.c,v 1.18 2008/09/18 05:42:08 dholland Exp $");
#endif
#endif /* not lint */

#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int, char *[]);

/* ARGSUSED */
int
main(int argc, char *argv[])
{
	int nflag;

	setprogname(argv[0]);
	(void)setlocale(LC_ALL, "");

	/* This utility may NOT do getopt(3) option parsing. */
	if (*++argv && !strcmp(*argv, "-n")) {
		++argv;
		nflag = 1;
	}
	else
		nflag = 0;

	while (*argv) {
		(void)printf("%s", *argv);
		if (*++argv)
			(void)putchar(' ');
	}
	if (nflag == 0)
		(void)putchar('\n');
	fflush(stdout);
	if (ferror(stdout))
		exit(1);
	exit(0);
	/* NOTREACHED */
}
